File: proxy/aero-l2-proxy/src/metrics.ts
const PING_RTT_MS_BUCKETS = [1, 5, 10, 25, 50, 100, 250, 500, 1000, 2500] as const;

function pushGauge(out: string[], name: string, val: number): void {
  out.push(`# TYPE ${name} gauge`);
  out.push(`${name} ${val}`);
}

function pushCounter(out: string[], name: string, val: bigint): void {
  out.push(`# TYPE ${name} counter`);
  out.push(`${name} ${val.toString()}`);
}

function pushZeroHistogram(out: string[], name: string, buckets: readonly number[]): void {
  out.push(`# TYPE ${name} histogram`);
  for (const bound of buckets) {
    out.push(`${name}_bucket{le="${bound}"} 0`);
  }
  out.push(`${name}_bucket{le="+Inf"} 0`);
  out.push(`${name}_sum 0`);
  out.push(`${name}_count 0`);
}

export class L2ProxyMetrics {
  private sessionsActive = 0;
  private sessionsTotal = 0n;

  private framesRxTotal = 0n;
  private framesTxTotal = 0n;
  private bytesRxTotal = 0n;
  private bytesTxTotal = 0n;
  private framesDroppedTotal = 0n;

  private policyDeniedTotal = 0n;

  // The quota/upgrade harness does not implement the full data plane, but we export the
  // production metric names so dashboards can be wired consistently.
  private tcpConnsActive = 0;
  private tcpConnectFailTotal = 0n;
  private udpFlowsActive = 0;
  private udpSendFailTotal = 0n;
  private dnsQueriesTotal = 0n;
  private dnsFailTotal = 0n;

  sessionOpened(): void {
    this.sessionsTotal += 1n;
    this.sessionsActive += 1;
  }

  sessionClosed(): void {
    if (this.sessionsActive > 0) this.sessionsActive -= 1;
  }

  frameRx(bytes: number): void {
    this.framesRxTotal += 1n;
    this.bytesRxTotal += BigInt(bytes);
  }

  frameTx(bytes: number): void {
    this.framesTxTotal += 1n;
    this.bytesTxTotal += BigInt(bytes);
  }

  frameDropped(): void {
    this.framesDroppedTotal += 1n;
  }

  policyDenied(): void {
    this.policyDeniedTotal += 1n;
  }

  renderPrometheus(): string {
    const out: string[] = [];

    pushGauge(out, "l2_sessions_active", this.sessionsActive);
    pushCounter(out, "l2_sessions_total", this.sessionsTotal);

    pushCounter(out, "l2_frames_rx_total", this.framesRxTotal);
    pushCounter(out, "l2_frames_tx_total", this.framesTxTotal);
    pushCounter(out, "l2_bytes_rx_total", this.bytesRxTotal);
    pushCounter(out, "l2_bytes_tx_total", this.bytesTxTotal);
    pushCounter(out, "l2_frames_dropped_total", this.framesDroppedTotal);

    pushCounter(out, "l2_policy_denied_total", this.policyDeniedTotal);

    pushGauge(out, "l2_tcp_conns_active", this.tcpConnsActive);
    pushCounter(out, "l2_tcp_connect_fail_total", this.tcpConnectFailTotal);
    pushGauge(out, "l2_udp_flows_active", this.udpFlowsActive);
    pushCounter(out, "l2_udp_send_fail_total", this.udpSendFailTotal);
    pushCounter(out, "l2_dns_queries_total", this.dnsQueriesTotal);
    pushCounter(out, "l2_dns_fail_total", this.dnsFailTotal);

    pushZeroHistogram(out, "l2_ping_rtt_ms", PING_RTT_MS_BUCKETS);

    return `${out.join("\n")}\n`;
  }
}

File: crates/aero-l2-proxy/src/metrics.rs
use std::sync::{
    atomic::{AtomicU64, Ordering},
    Arc,
};

#[derive(Clone)]
pub struct Metrics {
    inner: Arc<MetricsInner>,
}

struct MetricsInner {
    next_session_id: AtomicU64,

    // Sessions
    sessions_active: AtomicU64,
    sessions_total: AtomicU64,

    // Frames/bytes
    frames_rx_total: AtomicU64,
    frames_tx_total: AtomicU64,
    bytes_rx_total: AtomicU64,
    bytes_tx_total: AtomicU64,
    frames_dropped_total: AtomicU64,

    // Policy
    policy_denied_total: AtomicU64,

    // TCP proxy
    tcp_conns_active: AtomicU64,
    tcp_connect_fail_total: AtomicU64,

    // UDP proxy
    udp_flows_active: AtomicU64,
    udp_send_fail_total: AtomicU64,

    // DNS
    dns_queries_total: AtomicU64,
    dns_fail_total: AtomicU64,
}

impl Metrics {
    pub fn new() -> Self {
        Self {
            inner: Arc::new(MetricsInner {
                next_session_id: AtomicU64::new(1),
                sessions_active: AtomicU64::new(0),
                sessions_total: AtomicU64::new(0),
                frames_rx_total: AtomicU64::new(0),
                frames_tx_total: AtomicU64::new(0),
                bytes_rx_total: AtomicU64::new(0),
                bytes_tx_total: AtomicU64::new(0),
                frames_dropped_total: AtomicU64::new(0),
                policy_denied_total: AtomicU64::new(0),
                tcp_conns_active: AtomicU64::new(0),
                tcp_connect_fail_total: AtomicU64::new(0),
                udp_flows_active: AtomicU64::new(0),
                udp_send_fail_total: AtomicU64::new(0),
                dns_queries_total: AtomicU64::new(0),
                dns_fail_total: AtomicU64::new(0),
            }),
        }
    }

    pub fn next_session_id(&self) -> u64 {
        self.inner.next_session_id.fetch_add(1, Ordering::Relaxed)
    }

    pub fn session_opened(&self) {
        self.inner.sessions_total.fetch_add(1, Ordering::Relaxed);
        self.inner.sessions_active.fetch_add(1, Ordering::Relaxed);
    }

    pub fn session_closed(&self) {
        self.inner.sessions_active.fetch_sub(1, Ordering::Relaxed);
    }

    pub fn tcp_conn_opened(&self) {
        self.inner.tcp_conns_active.fetch_add(1, Ordering::Relaxed);
    }

    pub fn tcp_conn_closed(&self) {
        self.inner.tcp_conns_active.fetch_sub(1, Ordering::Relaxed);
    }

    pub fn tcp_connect_failed(&self) {
        self.inner
            .tcp_connect_fail_total
            .fetch_add(1, Ordering::Relaxed);
    }

    pub fn udp_flow_opened(&self) {
        self.inner.udp_flows_active.fetch_add(1, Ordering::Relaxed);
    }

    pub fn udp_flow_closed(&self) {
        self.inner.udp_flows_active.fetch_sub(1, Ordering::Relaxed);
    }

    pub fn udp_send_failed(&self) {
        self.inner
            .udp_send_fail_total
            .fetch_add(1, Ordering::Relaxed);
    }

    pub fn frame_rx(&self, bytes: usize) {
        self.inner.frames_rx_total.fetch_add(1, Ordering::Relaxed);
        self.inner
            .bytes_rx_total
            .fetch_add(bytes as u64, Ordering::Relaxed);
    }

    pub fn frame_tx(&self, bytes: usize) {
        self.inner.frames_tx_total.fetch_add(1, Ordering::Relaxed);
        self.inner
            .bytes_tx_total
            .fetch_add(bytes as u64, Ordering::Relaxed);
    }

    pub fn frame_dropped(&self) {
        self.inner
            .frames_dropped_total
            .fetch_add(1, Ordering::Relaxed);
    }

    pub fn policy_denied(&self) {
        self.inner
            .policy_denied_total
            .fetch_add(1, Ordering::Relaxed);
    }

    pub fn dns_query(&self) {
        self.inner.dns_queries_total.fetch_add(1, Ordering::Relaxed);
    }

    pub fn dns_fail(&self) {
        self.inner.dns_fail_total.fetch_add(1, Ordering::Relaxed);
    }

    pub fn render_prometheus(&self) -> String {
        let sessions_active = self.inner.sessions_active.load(Ordering::Relaxed);
        let sessions_total = self.inner.sessions_total.load(Ordering::Relaxed);
        let frames_rx_total = self.inner.frames_rx_total.load(Ordering::Relaxed);
        let frames_tx_total = self.inner.frames_tx_total.load(Ordering::Relaxed);
        let bytes_rx_total = self.inner.bytes_rx_total.load(Ordering::Relaxed);
        let bytes_tx_total = self.inner.bytes_tx_total.load(Ordering::Relaxed);
        let frames_dropped_total = self.inner.frames_dropped_total.load(Ordering::Relaxed);
        let policy_denied_total = self.inner.policy_denied_total.load(Ordering::Relaxed);
        let tcp_conns_active = self.inner.tcp_conns_active.load(Ordering::Relaxed);
        let tcp_connect_fail_total = self.inner.tcp_connect_fail_total.load(Ordering::Relaxed);
        let udp_flows_active = self.inner.udp_flows_active.load(Ordering::Relaxed);
        let udp_send_fail_total = self.inner.udp_send_fail_total.load(Ordering::Relaxed);
        let dns_queries_total = self.inner.dns_queries_total.load(Ordering::Relaxed);
        let dns_fail_total = self.inner.dns_fail_total.load(Ordering::Relaxed);

        let mut out = String::new();

        push_gauge(&mut out, "l2_sessions_active", sessions_active);
        push_counter(&mut out, "l2_sessions_total", sessions_total);

        push_counter(&mut out, "l2_frames_rx_total", frames_rx_total);
        push_counter(&mut out, "l2_frames_tx_total", frames_tx_total);
        push_counter(&mut out, "l2_bytes_rx_total", bytes_rx_total);
        push_counter(&mut out, "l2_bytes_tx_total", bytes_tx_total);
        push_counter(&mut out, "l2_frames_dropped_total", frames_dropped_total);

        push_counter(&mut out, "l2_policy_denied_total", policy_denied_total);

        push_gauge(&mut out, "l2_tcp_conns_active", tcp_conns_active);
        push_counter(
            &mut out,
            "l2_tcp_connect_fail_total",
            tcp_connect_fail_total,
        );

        push_gauge(&mut out, "l2_udp_flows_active", udp_flows_active);
        push_counter(&mut out, "l2_udp_send_fail_total", udp_send_fail_total);

        push_counter(&mut out, "l2_dns_queries_total", dns_queries_total);
        push_counter(&mut out, "l2_dns_fail_total", dns_fail_total);

        out
    }
}

fn push_gauge(out: &mut String, name: &str, val: u64) {
    out.push_str("# TYPE ");
    out.push_str(name);
    out.push_str(" gauge\n");
    out.push_str(name);
    out.push(' ');
    out.push_str(&val.to_string());
    out.push('\n');
}

fn push_counter(out: &mut String, name: &str, val: u64) {
    out.push_str("# TYPE ");
    out.push_str(name);
    out.push_str(" counter\n");
    out.push_str(name);
    out.push(' ');
    out.push_str(&val.to_string());
    out.push('\n');
}
